src/common/device_helpers.cuh

```diff
@@ -853,6 +857,8 @@ class AllReducer {
   std::vector<ncclComm_t> comms;
   std::vector<cudaStream_t> streams;
   std::vector<int> device_ordinals;  // device id from CUDA
+  std::vector<int> device_counts;  // device count from CUDA
```

Comment:
Making `device_counts` and `id` member variables means that the associated distributed operation for the `AllReducer` is fully defined. 

1. `id` is unique to a communicator clique, and can be used to differentiate between multiple `AllReducer`s in the cluster. 
2. `device_counts` defines a distributed `AllReducer`. In conjunction with (1), it can be used to capture failures, and introduce more robust fault tolerance.

When XGBoost used `ncclCommInitAll`, and the PIDs were identical, there was no reason to maintain `id`.

If you'd rather me remove these things until a future PR introducing distributed error handling, and fault tolerance, I think that would be fine, but I didn't see _harm_ from leaving this infrastructure in place to enable future PRs to leverage new error checking, etc.

Reply:
This is okay, as long as there is a purpose.